File: Leetcode/wordbreak.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

// 이건 wordDict fisrt임
 
bool wordBreak(string s, vector<string>& wordDict){
	int len = s.length();
	vector<int> cache(len+1, 0);
	cache[0] = 1;
	
	for(int i = 0; i < len; i++){
		if(cache[i] == 0) continue;
		for(auto& word : wordDict){
			int word_len = word.length();
			if(word_len <= len - i && s.substr(i, word_len) == word){
//				 만약에 대응하는 word가 있다면?
//				 i + word_len까지는 문장이 완성될 수 있다.
 				cache[i + word_len] = 1;
				 // 전체길이 + 1까지 와야 성공하는 것	
			}
		}
	}
	return cache[len];
}
int main(void){

	string s1 = "leetcode";
	vector<string> wd1 = {"leet", "code"};
	string s2 = "applepenapple";
	vector<string> wd2 = {"apple", "pen"};
	string s3 = "catsandog";
	vector<string> wd3 = {"cats", "dog", "sand", "and", "cat"};
	cout << wordBreak(s1, wd1) << "\n";
	cout << wordBreak(s2, wd2) << "\n";
	cout << wordBreak(s3, wd3) << "\n";
	return 0;
}

File: Leetcode/wordbreak.py

# wordDict first and string first 두 가지 관점으로 생각해볼수있다.
# wordDict first로 하면 명확하게 O(N**M)인데 나처럼 string first로 생각해보면
# 파이썬은 재귀가 너무 쌓여서 터질 수도 있음(결국 끝까지 가서 확인하고 return하기때문에)
class Solution:
    def wordBreak(self, s: str, wordDict) -> bool:
        cache = [-1 for _ in range(len(s))]

        def dfs(s: str, index):
            if cache[index] != -1:
                return cache[index]
            ret = False
            for i in range(0, len(s)):
                if s[0:i+1] in wordDict:
                    if i == len(s) - 1:
                        cache[index] = True
                        return cache[index]

                    ret = ret or dfs(s[i+1:], index + i + 1)

            cache[index] = ret
            return ret

        return dfs(s, 0)


s = Solution()
print(s.wordBreak("leetcode", ["leet", "code"]))
print(s.wordBreak("applepenapple", ["apple", "pen"]))
print(s.wordBreak("catsandog", ["cats", "dog", "sand", "and", "cat"]))
# s = "123123"
